blockprod/src/detail/mod.rs

```diff
@@ -47,158 +85,566 @@ impl BlockProduction {
+                            PropertyQueryError::GetAncestorError(
+                                GetAncestorError::InvalidAncestorHeight {
+                                    block_height: block_index.block_height(),
+                                    ancestor_height,
+                                },
+                            )
+                        })
+                    };
+
+                    let consensus_data = consensus::generate_consensus_data(
+                        &chain_config,
+                        &best_block_index,
+                        block_timestamp,
+                        best_block_index.block_height().next_height(),
+                        get_ancestor,
+                    );
+                    consensus_data.map(|cons_data| (cons_data, best_block_index))
                 }
-
-                None
             })
             .await?
-            .ok_or(BlockProductionError::FailedToConstructBlock(
-                BlockCreationError::CurrentTipRetrievalError,
-            ))?;
-
-        let (_tx, dummy_rx) = crossbeam_channel::unbounded();
-
-        let block_maker = BlockMaker::new(
-            self.chain_config.clone(),
-            self.chainstate_handle.clone(),
-            self.mempool_handle.clone(),
-            self.time_getter.clone(),
-            reward_destination,
-            current_tip_id,
+            .map_err(BlockProductionError::FailedConsensusInitialization)?;
+
+        Ok(consensus_data)
+    }
+
+    fn solve_block(
+        chain_config: Arc<ChainConfig>,
+        mut block_header: BlockHeader,
+        current_tip_height: BlockHeight,
+        stop_flag: Arc<AtomicBool>,
+    ) -> Result<BlockHeader, BlockProductionError> {
+        // TODO: use a separate executor for this loop to avoid starving tokio tasks
+        consensus::finalize_consensus_data(
+            &chain_config,
+            &mut block_header,
             current_tip_height,
-            dummy_rx,
-        );
+            stop_flag,
+        )?;
+
+        Ok(block_header)
+    }
 
-        Ok(block_maker.generate_block(transactions).await?)
+    pub fn make_job_key(tip_at_start: &GenBlockIndex) -> JobKey {
+        JobKey {
+            current_tip: tip_at_start.block_id(),
+        }
+    }
+
+    pub async fn generate_block(
+        &mut self,
+        _reward_destination: Destination,
+        transactions_source: TransactionsSource,
+    ) -> Result<Block, BlockProductionError> {
+        let stop_flag = Arc::new(false.into());
+
+        let (cancel_sender, mut cancel_receiver) = oneshot::channel::<()>();
+
+        let timestamp = BlockTimestamp::from_duration_since_epoch(self.time_getter().get_time());
+
+        let (_, tip_at_start) = self.pull_consensus_data(timestamp).await?;
+
+        {
+            // define the job and insert it into the map of all jobs
+            let job_key = Self::make_job_key(&tip_at_start);
+
+            #[allow(clippy::map_entry)]
```

Comment:
use `match self.all_jobs.entry(&job_key){...}`?

Reply:
It requires copying the job key. That's why I didn't do it.

Reply:
Using `Entry` should be the same because `VacantEntry::insert` doesn't require a key at all

Reply:
oh right, thanks